src/TreeSupport.h

```diff
@@ -7,10 +7,92 @@
 #include <forward_list>
 #include <unordered_set>
 
+#include <memory>
+#include <vector>
+#include <utility>
+#include <unordered_map>
+
 #include "sliceDataStorage.h"
 
+
 namespace cura
 {
+/*!
+ * \brief Lazily generates tree guidance volumes.
+ *
+ * NOTE: This class is not currently thread-safe and should not be accessed in OpenMP blocks
+ */
+class ModelVolumes
+{
+public:
+    ModelVolumes() = default;
+    ModelVolumes(const SliceDataStorage& storage, Polygons machine_border, coord_t xy_distance, coord_t max_move,
+                 coord_t radius_sample_resolution);
+
+    ModelVolumes(ModelVolumes&&) = default;
+    ModelVolumes& operator=(ModelVolumes&&) = default;
+
+    ModelVolumes(const ModelVolumes&) = delete;
+    ModelVolumes& operator=(const ModelVolumes&) = delete;
+
+    /*!
+     * \brief Creates the areas that have to be avoided by the tree's branches.
+     *
+     * The result is a 2D area that would cause nodes of radius \p radius to
+     * collide with the model.
+     *
+     * \param radius The radius of the node of interest
+     * \param layer The layer of interest
+     * \return Polygons object
+     */
+    const Polygons& collision(coord_t radius, int layer) const;
+
+    /*!
+     * \brief Creates the areas that have to be avoided by the tree's branches
+     * in order to reach the build plate.
+     *
+     * The result is a 2D area that would cause nodes of radius \p radius to
+     * collide with the model or be unable to reach the build platform.
+     *
+     * The input collision areas are inset by the maximum move distance and
+     * propagated upwards.
+     *
+     * \param radius The radius of the node of interest
+     * \param layer The layer of interest
+     * \return Polygons object
+     */
+    const Polygons& avoidance(coord_t radius, int layer) const;
+
+    /*!
+     * \brief Generates the area of a given layer that must be avoided if the
+     * branches wish to go towards the model
+     *
+     * The area represents the areas that do not collide with the model but
+     * are unable to reach the build platform
+     *
+     * \param radius The radius of the node of interest
+     * \param layer The layer of interest
+     * \return Polygons object
+     */
+    const Polygons& internal_model(coord_t radius, int layer) const;
+
+private:
+    using RadiusLayerPair = std::pair<coord_t, int>;
+
+    coord_t roundRadius(coord_t radius) const;
+    const Polygons& calculateCollision(const RadiusLayerPair& key) const;
+    const Polygons& calculateAvoidance(const RadiusLayerPair& key) const;
+    const Polygons& calculateInternalModel(const RadiusLayerPair& key) const;
+
+    Polygons machine_border_;
+    coord_t xy_distance_;
+    coord_t max_move_;
+    coord_t radius_sample_resolution_;
+    std::vector<Polygons> layer_outlines_;
+    mutable std::unordered_map<RadiusLayerPair, Polygons> collision_cache_;
```

Comment:
I think it's a shame you didn't implement these as `std::vector<std::vector<LazyInitialization<Polygons>>>`, but I find it a shame solely because I made that class and it has the word 'Lazy' in it ;)

Reply:
Using a vector of vectors would mean that a `LazyInitialization<Polygons>` would be allocated for every combination of radius and layer index. Even though the `Polygons` object is lazily allocated, there are still an unnecessary number of vectors allocated.
The unordered_map expresses the intent better to my mind.

Reply:
Yeah I know. I was just joking.

Reply:
Long day, humour detector faulty :)